docs: Add fentry/fexit sensor documentation (#5028)
cilium/tetragon: docs/content/en/docs/concepts/tracing-policy/hooks.md

    # [...]
```

## Fentry/Fexit

Fentry (BPF fentry/fexit) programs attach to kernel functions using BTF (BPF Type Format). They

> are a modern alternative to kprobes with lower overhead. Instead of using breakpoints and

Reviewer: let's just state trampolines are faster alternative to kprobes, the breakpoint part is bit more difficult and not really accurate, also safer is arguable

> Fentry (BPF fentry/fexit) programs attach to kernel functions using BTF (BPF Type Format). They
> are a modern alternative to kprobes with lower overhead. Instead of using breakpoints and
> instruction patching like kprobes, fentry programs are called directly from a trampoline, which
> makes them faster and safer.

Reviewer 2: ```suggestion
Fentry/fexit programs attach to kernel functions using BTF (BPF Type Format).
They are a modern alternative to kprobes with lower overhead. Instead of using
breakpoints and instruction patching like kprobes, fentry programs rely on the
ftrace kernel infrastructure which calls the BPF progs directly from a
trampoline making them faster and safer.
```

### When to use fentry vs kprobes

Fentry is preferred over kprobes when the environment supports it, because:

- Lower overhead: no breakpoint or instruction patching is needed; the BPF program is invoked
  directly from a trampoline.

> - Safer: avoids the complexity and potential issues of runtime instruction patching.

Reviewer: arguable.. kprobes on function entries use ftrace in a similar way like trampolines,
maybe you mean the kprobe attach complexity when it's attached in the middle of the function
on specific instruction, than it's tricky.. but still safe

However, kprobes remain the better choice when:

- Older kernels: fentry in Tetragon requires kernel >= 6.1.
- No BTF support: fentry requires the kernel to be built with `CONFIG_DEBUG_INFO_BTF=y`.
- Enforcement is needed: fentry does not yet support enforcement (override) actions.

### Kernel requirements

To use fentry, the following kernel requirements must be met:

- Kernel version: >= 6.1 is required for Tetragon's fentry implementation.
- BTF support: the kernel must be built with BTF enabled (`CONFIG_DEBUG_INFO_BTF=y`). You can

>   verify this with:

Reviewer 2: I feel like we kinda repeat ourselves at a few lines of distance

```shell
cat /boot/config-$(uname -r) | grep CONFIG_DEBUG_INFO_BTF
```

The output should include:

```
CONFIG_DEBUG_INFO_BTF=y
```

### Limitations

- No enforcement/override support: unlike kprobes, fentry does not yet support enforcement

>   actions such as `Override` or `Signal`.

Reviewer 2: This as well is repeated

- No multi-attach support: each fentry spec attaches to a single kernel function.

### Example

Here is an example of a `TracingPolicy` using the `fentries` section to trace the `tcp_connect`
kernel function:

```yaml
apiVersion: cilium.io/v1alpha1
kind: TracingPolicy
metadata:
  name: "fentry-example"
spec:
  fentries:
  - call: "tcp_connect"
    syscall: false
    return: true
    args:
    - index: 0
      type: "sock"
    returnArg:
      index: 0
      type: "int"
```

{{< note >}}
Unlike kprobes where `syscall` can be set to `true` to trace system call entry points, fentry is
typically used for internal kernel functions and `syscall` should be set to `false`. This is
because fentry attaches via BTF-based trampolines that are suited for tracing internal kernel
function boundaries rather than the syscall ABI layer.

> {{< /note >}}

Reviewer 2: question: can you trace syscall with fentry? it's not clear to me when reading this

Reviewer: yes, the `__x64_sys_*` are traceable and need special processing to get the syscall arguments,
so you need to use the syscall flag for fentry sensor as well

{{< note >}}

> The `fentry` program attaches at function *entry*, while its `fexit` counterpart attaches at

Reviewer 3: Maybe I'm missing it, but I can't see a corresponding explanation for kretprobes; I think we just silently hide the kretprobe functionality behind capturing the return value in a kprobe. Therefore, I question whether we need this discussion about fexit here, as the same could apply to fentry?

> The `fentry` program attaches at function *entry*, while its `fexit` counterpart attaches at
> function *exit*, where the return value is available. To capture a function's return value, set

Reviewer 2: https://kubernetes.io/docs/contribute/style/style-guide/#use-italics-to-define-or-introduce-new-terms
```suggestion
The `fentry` program attaches at function entry, while its `fexit` counterpart attaches at
function exit, where the return value is available. To capture a function's return value, set
```

> {{< note >}}
> The `fentry` program attaches at function *entry*, while its `fexit` counterpart attaches at
> function *exit*, where the return value is available. To capture a function's return value, set
> `return: true` — this is what attaches the `fexit` program — and define `returnArg` with the
> return value's `type`, as shown in the example above. When `return` is left unset (it defaults to
> `false`), only the `fentry` program is attached and any `returnArg` is ignored. See
> [Return values](#return-values) for more details on `return` and `returnArg`.
> {{< /note >}}

Reviewer 2: My opinion but I feel like it could be a normal paragraph, looks like important information to explain how to do fexit with the tracingpolicy!
```suggestion
The `fentry` program attaches at function *entry*, while its `fexit` counterpart attaches at
function *exit*, where the return value is available. To capture a function's return value, set
`return: true` — this is what attaches the `fexit` program — and define `returnArg` with the
return value's `type`, as shown in the example above. When `return` is left unset (it defaults to
`false`), only the `fentry` program is attached and any `returnArg` is ignored. See
[Return values](#return-values) for more details on `return` and `returnArg`.
```
